### PR TITLE
BUG: change poly() real output checking to test if all imaginary parts are zero

### PR DESCRIPTION
**Pull request by endolith**
Some arrays should produce real output but don't:

```
In [144]: poly([1j, -1j])
Out[144]: array([ 1.,  0.,  1.])

In [145]: poly([1j, -1j, 2j, -2j])
Out[145]: array([ 1.+0.j,  0.+0.j,  5.+0.j,  0.+0.j,  4.+0.j])

In [146]: _.imag == 0
Out[146]: array([ True,  True,  True,  True,  True], dtype=bool)
```

I think the check was meant to be `sort_complex(conjugate(...` rather than `conjugate(sort_complex(...`, which would work, but checking that the imaginary parts are exactly zero is equivalent and simpler anyway.

### CODE REVIEW

File: numpy/lib/polynomial.py [hunk 1]
```diff
@@ -139,12 +139,11 @@ def poly(seq_of_zeros):
     if issubclass(a.dtype.type, NX.complexfloating):
         # if complex roots are all complex conjugates, the roots are real.
         roots = NX.asarray(seq_of_zeros, complex)
-        pos_roots = sort_complex(NX.compress(roots.imag > 0, roots))
-        neg_roots = NX.conjugate(sort_complex(
-                                        NX.compress(roots.imag < 0, roots)))
-        if (len(pos_roots) == len(neg_roots) and
-            NX.alltrue(neg_roots == pos_roots)):
-            a = a.real.copy()
+        pos_roots = NX.compress(roots.imag > 0, roots)
```
Comment (charris) [inline numpy/lib/polynomial.py]: I think you could just check `np.all(roots.sort() == roots.conjugate().sort())`.

Comment (endolith) [inline numpy/lib/polynomial.py]: Well, `.sort()` is in-place sorting and doesn't return anything, so `anything.sort() == anythingelse.sort()` is always true.

`np.all(np.sort(roots) == np.sort(roots.conjugate))`?

`sort_complex()` just calls `sort()` internally but always outputs a complex type?  Not sure if it's necessary, but just kept what was in the original code.

Comment (charris) [inline numpy/lib/polynomial.py]: You're right about the sort method.

Both the polynomial and sort_complex code date back to 2002 with later refactoring, so I expect not using sort directly is just historical accident.

Comment (charris) [inline numpy/lib/polynomial.py]: And since only complex types are checked, the return will be complex anyway.

Comment (endolith) [inline numpy/lib/polynomial.py]: I assume checking the size first is to speed up rejections for huge arrays?

Comment (endolith) [inline numpy/lib/polynomial.py]: Though I guess polynomials are always going to be short 1d arrays?

Comment (charris) [inline numpy/lib/polynomial.py]: I'd guess that going much over 20 is going to be a problem in general. I've gone up to 60 equally spaced roots in the polynomial package, which does better than this, but the results can be sensitive to root placement.

**Comment by jaimefrio** [comment]
You need to rebase on master, some changes were made to the type handling a while back.

On a separate note regarding the actual object of this PR, wouldn't it make sense to detect the conjugate roots **before** multiplying? That way you could collapse every two conjugate roots into a strictly real second degree binomial and avoid roundoff errors creeping into the imaginary part of the result.

**Comment by charris** [comment]
@jaimefrio Yes, and for power series that is pretty easy to do, although the convolution here would need modification. For other polynomial basis it gets a bit more involved.

**Comment by endolith** [comment]
Ok, I rebased it and fixed an import error.  Should I definitely change to `np.all(np.sort(roots) == np.sort(roots.conjugate()))`?

**Comment by charris** [comment]
@endolith I think it would be cleaner. I don't think speed is of much  importance here, but it might even be faster.

**Comment by endolith** [comment]
Ok I changed it.

**Comment by rgommers** [comment]
@charris this seems about ready to merge. ignore the test failure, that's unrelated (couldn't download Cython)

**Comment by endolith** [comment]
Actually instead of that long string of numbers in the test, it could just set the numpy random seed for determinism and go back to 
```
a = rand(100) + 1j* rand(100)
assert_(np.isrealobj(np.poly(np.concatenate((a, np.conjugate(a))))))
```

**Comment by endolith** [comment]
Was there anything else that needed to be done on this?

**Comment by endolith** [comment]
Is there anything wrong with this change?

**Comment by charris** [comment]
No, LGTM ;) Thanks @endolith .

**Pull request merged**